{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installation notes \n",
    "\n",
    "##### Prerequisites \n",
    "Python v3 \\\n",
    "Jupyter notebook \n",
    "\n",
    "##### Python packages to run the code\n",
    "pip install networkx \\\n",
    "pip install matplotlib "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gfa parser to write the gfa file to dictiory and add repeats after confirmation\n",
    "#the first step in the code \n",
    "from collections import defaultdict\n",
    "\n",
    "class gfa_parser:\n",
    "    '''\n",
    "    Read the spades output in GFA v1.0 graph to genrate \n",
    "    nodes and edges. \n",
    "    \n",
    "    #Class initiation\n",
    "    x=gfa_parser(\"chrom2.gfa\")\n",
    "    \n",
    "    #calling a specific function within the class\n",
    "    print(x.gfa_parser(\"chrom2.gfa\"))\n",
    "    #prints all the nodes and edges from the gfa \n",
    "    \n",
    "    '''\n",
    "    def __init__(self, gfa):\n",
    "        self.gfa=gfa\n",
    "        \n",
    "    \n",
    "    def gfa_parser(self, gfa):\n",
    "        #reading the gfa file to a new variable\n",
    "        g1=open(gfa, \"r\")\n",
    "        nodes=[]\n",
    "        edges=[]\n",
    "        coverage={}\n",
    "        sequences={}\n",
    "        for line in g1:\n",
    "            if (line.startswith(\"L\")):\n",
    "                link=line.split('\\t')\n",
    "                edges.append([link[1],link[3]])\n",
    "            \n",
    "            elif (line.startswith(\"S\")):\n",
    "                info=line.split('\\t')\n",
    "                seq=float(info[1])\n",
    "                sequences[seq]=info[2]\n",
    "                cov=info[-1].split(\":\")[-1]\n",
    "                cov=float(cov.strip(\"\\n\"))\n",
    "                if (cov>2.0):\n",
    "                    coverage[info[1]]=cov\n",
    "        \n",
    "        #user input\n",
    "        repeats=[]\n",
    "        for f in repeats:\n",
    "            cv=coverage.get(f)\n",
    "            cv=cv-1\n",
    "            temp=f\n",
    "            for index in range (int(cv)):\n",
    "                rename=str(f)+\"r\"+str(index)\n",
    "                edges.append([temp,rename])\n",
    "                temp=rename\n",
    "                \n",
    "        return(edges, coverage, sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes\n",
    "\n",
    "***Vibrio splendidus* genome Chromosome 1 contains 7 rRNA genes and 97 tRNA genes, while chromosome 2 contains 1 rRNA gene and 17 tRNA genes**\n",
    "\n",
    "\n",
    "\n",
    "**Chromsome 2 high coverage regions**\n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after| Annotation |\n",
    "|---|-----|-----|---------|-----|-----|\n",
    "|13|3.10| 410bp | 1| 1| -|\n",
    "|19|7.77| 249bp| 1| 1| -|\n",
    "|20|5.64| 247bp| 1| 1| -|\n",
    "|22|5.74| 204bp| 2| 2|-|\n",
    "|27|4.18| 162bp| 1| 2|-|\n",
    "|79|3.25| 1bp|-| -| |\n",
    "|81|3.00| 1bp|-| -| | \n",
    "\n",
    "Conclusion - no more changes added. \n",
    "Running the same steps for chrom1\n",
    "\n",
    "**Chromosome 1 high coverage regions**\n",
    "\n",
    "rrn operons \n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after|Annotation|\n",
    "|---|-----|-----|---------|-----|-----|\n",
    "|14 |4.89 | 359bp| 16| 22| 5S ribosomal RNA|\n",
    "|30 |6.76 | 129bp| 8| 11| no associated gene/protein|\n",
    "|47 |5.02 | 68bp| 6| 9| hypothetical protein|\n",
    "|50 |5.95 | 58bp| 6| 9| 16S ribosomal partial protein|\n",
    "|10 |5.34 |1407bp | 4| 6| 16S ribosomal sequence|\n",
    "|17 |2.32 | 279bp| 10| 13|16S-23S ribosomal RNA intergenic spacer,|\n",
    "|18 |3.62 | 273bp| 11| 13| tRNA|\n",
    "|35 |4.28 | 108bp| 9| 12| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|67 |3.34 | 73bp| 7| 9| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|66 |2.41 | 73bp| 7| 9| hypothetical protein|\n",
    "|8 |11.39 | 2947bp|4 | 6| 23S ribosomal sequence|\n",
    "|39 |4.89 | 100bp| 7| 8| 5S ribosomal sequence|\n",
    "|33 |7.14 | 110bp| 9| 10| 5S ribosomal sequence|\n",
    "|41 |3.31 | 91bp| 7| 8| 5S ribosomal sequence|\n",
    "|59 |4.89 | 36bp| 6| 8| 5S ribosomal sequence|\n",
    "|60 |9.09 | 36bp| 6| 8| 5S ribosomal seuquence|\n",
    "|26 |6.29 | 163bp| 7| 10| 5S ribosomal sequence|\n",
    "|28 |2.41 | 157bp| 8| 11| 5S ribosomal sequence|\n",
    "|34 |2.03 | 109bp| 9| 12| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|44 |8.25 | 73bp| 5| 6| 16S-23S ribosomal RNA intergenic spacer |\n",
    "|45 |2.97 | 73bp| 5| 7| 16S ribosomal partial sequence|\n",
    "|52 |2.29 | 55bp| 6| 7| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|53 |3.30 | 53bp| 6| 9| no associated gene/protein|\n",
    "|55 |3.11 | 44bp| 1| 2| no associated gene/protein|\n",
    "|57 |4.25 | 39bp| 5| 6| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|58 |2.86 | 38bp| 6| 8| 16S-23S ribosomal RNA intergenic spacer|\n",
    "\n",
    "The below sequences are repats that aligned multiple times away from the rrn operon. \n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after|Annotation|\n",
    "|---|-----|-----|---------|-----|-----|\n",
    "|16 |3.38 | 285bp| 2| 2| no hits|\n",
    "|23 |3.25 | 189bp| 2| 2| no hits|\n",
    "|25 |3.00 | 174bp| 4| 5| tRNA|\n",
    "|31 |4.76 | 128bp| 1| 1| no hits |\n",
    "|40 |4.88 | 95bp| 1| 2| 16S-23S ribosomal RNA intergenic spacer,|\n",
    "|63 |2.84 | 29bp| 2| 3| no associated gene/protein|\n",
    "\n",
    "These are the really short sequences \n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after|\n",
    "|---|-----|-----|---------|-----|\n",
    "|69 |5.27 | 18bp| 0| 0| \n",
    "|71 |8.19 | 11bp| 0| 0| \n",
    "|72 |35.88 | 8bp| 0| 0| \n",
    "|73 |2.32 | 8bp| 0| 0| \n",
    "|74 |2.89 | 6bp| 0| 0| \n",
    "|75 |5.72 | 5bp| 0| 0| \n",
    "|77 |3.29 |2bp | 0| 0| \n",
    "|78 |3.43 |2bp | 0| 0| \n",
    "|80 |5.77 | 1bp| 0| 0| \n",
    "\n",
    "\n",
    "This genome has 5 complete (1 inocomplete) rrn sequences in chromosome1 and 1 complete rrn in chrosome2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python3 program to find the maximum depth of tree from a given node\n",
    "# counts the depth of the tree from the left tree and right tree\n",
    "# Python3 program to find the maximum depth of tree \n",
    "import networkx as nx\n",
    "# A binary tree node \n",
    "def maxDepth (node, dfs, graph): \n",
    "    if node is None: \n",
    "        return (0)\n",
    "    else : \n",
    "        # Compute the depth of each subtree \n",
    "        n=list(G.nodes())[0]\n",
    "        value=list(dfs[node])\n",
    "        left_tree=0\n",
    "        right_tree=0\n",
    "        left=value[0]\n",
    "        right=value[1]\n",
    "        while (left is not None):\n",
    "            try:\n",
    "                left=dfs[left][0]\n",
    "                left_tree=left_tree+1\n",
    "            except KeyError:\n",
    "                break\n",
    "        while (right is not None):\n",
    "            try:\n",
    "                right=dfs[right][0]\n",
    "                right_tree=right_tree+1\n",
    "            except KeyError:\n",
    "                break\n",
    "        \n",
    "    return (left_tree+1, right_tree+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if the node has two branches\n",
    "# 1) check the depth of the right and the left branch\n",
    "# 2) traverse through the branch that is the shortest then traverse back to the node that had two branches back to the next branch\n",
    "# 3) also checks to make sure the first node remains to have only 1 neighbor\n",
    "def adding_nodes_with_two_branches(key, successors, old_graph, new_graph):\n",
    "    d=maxDepth(key, successors, old_graph)\n",
    "    left_depth=d[0]\n",
    "    right_depth=d[1]\n",
    "    \n",
    "    #if the maximum depth to the left tree and right tree are both 1\n",
    "    if (right_depth==1) and (left_depth==1):\n",
    "        new_graph=check_before_adding(key, val[0], new_graph)\n",
    "        #getting the last node\n",
    "        last_node=get_last_edge(new_graph)[1]\n",
    "        #tracking back\n",
    "        new_graph=check_before_adding(last_node, key, new_graph)\n",
    "        new_graph=check_before_adding(key, val[1], new_graph)\n",
    "                    \n",
    "    #if the left depth is less than the right depth\n",
    "    elif (left_depth<right_depth):    \n",
    "        #print (\"left depth is less than right depth\", left_depth, right_depth)\n",
    "        if (left_depth==1):\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "            #getting the last node\n",
    "            last_node=get_last_edge(new_graph)[1]\n",
    "            #tracing back\n",
    "            new_graph=check_before_adding(last_node, key, new_graph)\n",
    "            new_graph=check_before_adding(key, val[1], new_graph)\n",
    "        else:\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "            successors[key]=(val[1])\n",
    "\n",
    "    #if the right depth is less than the left depth\n",
    "    elif (left_depth>right_depth):\n",
    "        #print (\"left depth is more than right depth\", left_depth, right_depth)\n",
    "        if (right_depth==1):\n",
    "            new_graph=check_before_adding(key, val[1], new_graph)\n",
    "            #getting the last node\n",
    "            last_node=get_last_edge(new_graph)[1]\n",
    "            #tracing back\n",
    "            new_graph=check_before_adding(last_node, key, new_graph)\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "        else:\n",
    "            new_graph=check_before_adding(key,val[1], new_graph)\n",
    "            successors[key]=(val[0])\n",
    "            \n",
    "    return (successors, new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the last edge and nodes added to the new tree\n",
    "def get_last_edge(new_graph):\n",
    "    edges=list(new_graph.edges())\n",
    "    last_edge=edges[-1]\n",
    "    last_node=last_edge[1].split('_')[0]\n",
    "    prev_node=last_edge[0].split('_')[0]\n",
    "    return (last_edge, last_node, prev_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function only renames the node since it already exists in the \n",
    "#   new graph \n",
    "# assumes that the node was already checked to confirm that \n",
    "#   the node is already added to the new graph \n",
    "def rename_node(node, new_graph):\n",
    "    nodes=new_graph.nodes()\n",
    "    count=1\n",
    "    new_node=node+\"_\"+str(count)\n",
    "    if (new_node in nodes):\n",
    "        ngbr=len(list(new_graph.neighbors(new_node)))\n",
    "        if (ngbr==1):\n",
    "            new_node=new_node\n",
    "        elif (ngbr==2):\n",
    "            count=count+1\n",
    "            new_node=node+\"_\"+str(count)\n",
    "    \n",
    "    return (new_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#before adding a  new edge, check if the nodes in the edge are \n",
    "#  1) first that the node exists or not in the new_graph \n",
    "#  2) check if the node neighbors are 2 or 1\n",
    "#  3) node1 and node2 is not the first node in the graph\n",
    "def check_before_adding(node1, node2, new_graph):\n",
    "    \n",
    "    #getting the first node in the new graph\n",
    "    n=list(G.nodes())[0]\n",
    "        \n",
    "    if (node1 not in new_graph.nodes()) and (node2 not in new_graph.nodes()):\n",
    "        new_graph.add_edge(node1, node2)\n",
    "\n",
    "    elif (node1 in new_graph.nodes()) and (node2 not in new_graph.nodes()):\n",
    "        n1_ngbrs=len(list(new_graph.neighbors(node1)))\n",
    "        if (n1_ngbrs==1) and (node1 !=n):\n",
    "            new_graph.add_edge(node1, node2)\n",
    "        elif (n1_ngbrs==1) and (node1 ==n):\n",
    "            x=rename_node(node1, new_graph)\n",
    "            new_graph.add_edge(x, node2)\n",
    "        elif (n1_ngbrs==2):\n",
    "            x=rename_node(node1, new_graph)\n",
    "            new_graph.add_edge(x, node2)\n",
    "\n",
    "    elif(node1 not in new_graph.nodes()) and (node2 in new_graph.nodes()):\n",
    "        n2_ngbrs=len(list(new_graph.neighbors(node2)))\n",
    "        if (n2_ngbrs==1) and (node2 !=n):\n",
    "            new_graph.add_edge(node1, node2)\n",
    "        elif (n2_ngbrs==1) and (node2 ==n):\n",
    "            x=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(node1, x)\n",
    "        elif (n2_ngbrs==2):\n",
    "            x=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(node1, x)\n",
    "\n",
    "    elif (node1 in new_graph.nodes()) and (node2 in new_graph.nodes()):\n",
    "        n1_ngbrs=len(list(new_graph.neighbors(node1)))\n",
    "        n2_ngbrs=len(list(new_graph.neighbors(node2)))\n",
    "        if (n1_ngbrs==1) and (n2_ngbrs==1):\n",
    "            if (node1 ==n) and (node2 !=n):\n",
    "                x=rename_node(node1, new_graph)\n",
    "                new_graph.add_edge(x,node2)\n",
    "            elif (node1 !=n) and (node2 ==n):\n",
    "                x=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(node1, x)\n",
    "            else:\n",
    "                new_graph.add_edge(node1, node2)\n",
    "\n",
    "        elif (n1_ngbrs==2) and (n2_ngbrs==1):\n",
    "            if (node2 !=n):\n",
    "                x=rename_node(node1, new_graph)\n",
    "                new_graph.add_edge(x, node2)\n",
    "            else:\n",
    "                k=rename_node(node1, new_graph)\n",
    "                v=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(k,v)\n",
    "    \n",
    "        elif (n1_ngbrs==1) and (n2_ngbrs==2):\n",
    "            if (node1 !=n):\n",
    "                x=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(node1,x)\n",
    "            else:\n",
    "                k=rename_node(node1, new_graph)\n",
    "                v=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(k,v)\n",
    "        else:\n",
    "            k=rename_node(node1, new_graph)\n",
    "            v=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(k,v)\n",
    "    return(new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finding the paths from the last node to the first node and selecting the \n",
    "# shortest path\n",
    "# using the dictionary to retrieve the neighbors\n",
    "def path_to_first_node(last_node, first_node, old_graph):\n",
    "    #getting the neighbors of the last node and working on adding the nodes to the graph accordingly\n",
    "    last_node=last_node.split(\"_\")[0]\n",
    "    last_node=last_node.split(\"r\")[0]\n",
    "    nbrs=list(old_graph.neighbors(last_node))\n",
    "    left_path=list(nx.shortest_path(old_graph, source=nbrs[0], target=first_node))\n",
    "    right_path=list(nx.shortest_path(old_graph, source=nbrs[1], target=first_node))\n",
    "\n",
    "    return (left_path, right_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#traversing through the two shortest paths from the last node to the key \n",
    "def shortest_path_to_first_node(left_path, right_path, new_graph):\n",
    "                                \n",
    "    #getting the last_node added to the new_graph \n",
    "    last_node=list(new_graph.nodes())[-1]\n",
    "    \n",
    "    #traversing the shortest path back\n",
    "    if (len(left_path)==len(right_path)):\n",
    "        for f in range(len(left_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, left_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(left_path[f-1], left_path[f], new_graph)\n",
    "                \n",
    "    elif (len(left_path)<len(right_path)):\n",
    "        for f in range(len(left_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, left_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(left_path[f-1], left_path[f], new_graph)\n",
    "                \n",
    "    elif (len(left_path)>len(right_path)):\n",
    "        for f in range(len(right_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, right_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(right_path[f-1], right_path[f], new_graph)\n",
    "    \n",
    "    return (new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writing out the complete graph as a fasta file. \n",
    "# the new complete graph is written out to \"new_graph\" variable, \n",
    "# but this holds only inforamtion about the node names and the sequence itself.\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def graph_to_fasta(graph, name, sequences):\n",
    "    #made sure the nodes were added in the correct order\n",
    "    nodes=graph.nodes()\n",
    "    final_seq=\"\"\n",
    "    length=len(nodes)-1\n",
    "    count=0\n",
    "    for f in nodes:\n",
    "        if (count<length):\n",
    "            raw_node=f.split(\"_\")[0]\n",
    "            raw_node=raw_node.split(\"r\")[0]\n",
    "            #print (raw_node)\n",
    "            seq=sequences[int(raw_node)]\n",
    "            final_seq = final_seq+seq\n",
    "            count=count+1\n",
    "    final_seq=Seq(final_seq)\n",
    "    \n",
    "    #writing out the final seq as a record class\n",
    "    final = SeqRecord(final_seq, id=name, description=\"putting it together\")\n",
    "    return (final)\n",
    "    \n",
    "###############################################\n",
    "# Driver program\n",
    "##############################################\n",
    "#final=graph_to_fasta(new_graph, sequences)\n",
    "#print (final.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'9': ['65'], '65': ['1'], '1': ['19'], '19': ['13'], '13': ['81'], '81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 9 ['65'] 0\n",
      "successors= {'65': ['1'], '1': ['19'], '19': ['13'], '13': ['81'], '81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 65 ['1'] 1\n",
      "successors= {'1': ['19'], '19': ['13'], '13': ['81'], '81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 1 ['19'] 2\n",
      "successors= {'19': ['13'], '13': ['81'], '81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 19 ['13'] 3\n",
      "successors= {'13': ['81'], '81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 13 ['81'] 4\n",
      "successors= {'81': ['22'], '22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 81 ['22'] 5\n",
      "successors= {'22': ['79'], '79': ['27'], '27': ['20']}\n",
      "key, val, count 22 ['79'] 6\n",
      "successors= {'79': ['27'], '27': ['20']}\n",
      "key, val, count 79 ['27'] 7\n",
      "successors= {'27': ['20']}\n",
      "key, val, count 27 ['20'] 8\n",
      "successors= {}\n",
      "[('9', '65'), ('65', '1'), ('1', '19'), ('19', '13'), ('13', '81'), ('81', '22'), ('22', '79'), ('79', '27'), ('27', '20'), ('20', '27_1'), ('27_1', '79_1'), ('79_1', '9_1')]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb4AAAEuCAYAAADx63eqAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAwTElEQVR4nO3de1xUdcIG8GcuwJAyEkqJopKiDLmJga1umg61VuJbeSVNU7uYrparJrvu8q7VvtFWg5mmZqvuZuKFYruvS5kyiAmCaLilo2LaSooCSsMIM87lvH8QbKSizJzhzMx5vv/YR5kfD+7Kw+/M76IQBEEAERGRTCilDkBERNSeWHxERCQrLD4iIpIVFh8REckKi4+IiGSFxUdERLLC4iMiIllh8RERkayw+IiISFZYfEREJCtqqQMQkWeqLTbklFbAVGmG2eqAVqOGrqsWE5Oi0bljiNTxiHyOgmd1EvmnslO1WGUsR/7RKgCAzeFq/jONWgkBgD4uEnNGxCKhR7g0IYl8EIuPyA9lFZ1ExjYTrA4nWvsXrFAAGrUK6Sk6TB0S0275iHwZH3US+ZnG0juMBrvrmh8rCECD3YmMbYcBgOVHBC5uIfIrZadqkbHNdF2l91MNdhcytplwsKLWO8GI/AiLj8iPrDKWw+pwuvVaq8OJ1cZykRMR+R8WH5GfqLbYkH+0qtX39FojCEDekSrUWGziBiPyM3yPj0TDZfXelVNa4fEYCgA5+yswa3gfzwMR+SkWH3ms9WX1lVj2xVEuqxeBqdLc4u/WHVaHC6YzdSIlIvJPLD7yyLWW1Vt//Eb9+aGz2HW0msvqf+bSpUuorq5GVVUVqqqqmv/7579WVVXhh4TJUPUc6PHnNFvtngcn8mMsPnIbl9W3JAgCzGbzNQvsp79XX1+Pzp07IzIyEl26dEFkZGTzf996663Nv9elSxes3n8Rnx294HHO2nOnUV1djS5duojwVRP5HxYfucXTZfUDosMxIDrcO+FEYrfbUVNT06YiCwkJaVFiP/01Li7usnILDw+HQqG4rjy31x6H8dsfPHrcqYILFV8Xo0+faejVqxf0ej30ej1GjBiBzp07uz0ukT/hyS3klqc27sP2w2fdWmGoUAD33Xoz1kwdJH6wqxAEARaL5boLrKqqCnV1dYiIiLhikV2t3DQajde+hmqLDUNf2elR8YWoldjz+7vRSaPC/v37YTQaYTQa8eWXXyImJqZFEUZERIiYnsh3sPiozcT8Buzuak+n04mampo2FZlKpbpmcf30z2688UYolb6148dbP3A4HA6Ulpa2KMLevXs3F+Hw4cNZhBQwWHzUZmvyj2PZF0c9Kj6NWokFI/s1L6uvr6+/7gKrrq5GbW0twsPDr6vAmn694YYbxPorkEzZqVpMWluEBnvbN7GHBqmQ/dSQ63rEbLfbWxThnj170KdPnxZFeOONN7rxFRBJj8VHbTY/+wA+/Oq0x+OEnP4KVuNfUVVVBZfLdVlZtVZqERERUKlUInw1/qcti4qahAYpkZ4S7/aiIrvdjn379jUXYWFhIWJjY6HX65GcnIy77roL4eHhbo1N1N5YfNRmj28owU7TOY/HSeoajKUP9UWXLl3QoUOH617kQdLfznDp0qXLirBfv34tirBTp06ifT4iMbH4qM3EmvGNHdgdyx4e6HkgmTpYUYvVxnLkHamCAv/dMwn89z6+5LhIzNHHen0F7aVLl1BSUtJchEVFRYiLi2suwmHDhrEIyWew+KjNvPEeH7mvxmJDzv4KmM7UwWy1Q6sJgi4qDBMSpTsqzmaztSjCvXv3QqfTtShCrVYrSTYiFh+1icPhwDvZ7+PPZcGAKsjtcTxd1Un+xWazobi4uLkIi4uLER8f36IIw8LCpI5JMsHio+ty8eJF/P3vf8drr72GqKgoRDy0GF/XKv1mHx/5FqvVelkR9u/fv7kIhw4dyiIkr2HxUavOnTuHlStXYs2aNRg6dCjS0tJw5513ttuyepIHq9WKvXv3NhdhSUkJfvGLX7Qowo4dO0odkwIEi4+u6NixY1i6dCmys7ORmpqKZ599Fv369WvxMVIsqyd5sFqtKCoqgtFoRF5eHkpLS3Hbbbc1F+Gdd97ZrkXIK7cCC4uPWigsLITBYEBBQQFmz56Np59+GjfffPNVP17qZfUkDw0NDS2KcP/+/RgwYECLIuzQoYPon7f1K7caV87yyi3/w+IjuFwufPzxxzAYDDh9+jQWLlyIxx9//Lq/kfjSsnqSh/r6+hZFeODAASQkJCA5ORl6vR533nmnxyf18Ie6wMXikzGr1Yp33nkHS5cuhVarRVpaGsaNGwe12r1LO3xxWT3JQ319PQoLC5uL8KuvvsLAgQObi/BXv/pVm4qQj/EDG4tPhs6fP4/Vq1dj5cqVSEpKQlpaGkaMGMGTUyhgXLx4sUURlpWV4fbbb29RhKGhoVd8LRduBT4Wn4ycOHECy5YtQ1ZWFh566CEsWrQI/fv3lzoWkdddvHgRe/bsaS7CgwcPIjExsbkIhwwZ0lyE/nblFrUdi08GSktLYTAYsH37djz55JOYN28eunfvLnUsIslYLJYWRfjvf/8bSUlJGKIfifcdt6ON9yu3wMMZfB+LL0AJgoDc3FwYDAYcO3YM8+fPx8yZM3lMFNEVWCwWfPnll1htPIYyZ7RHpxLxOD7f594qBvJZly5dwpYtW5CZmQmFQoG0tDRMmjQJQUHu/0MmCnQdO3bEfffdh3/V3oQyDw9gtzpcMJ2pEykZeQOLL0D88MMP+Otf/4rly5cjPj4emZmZuPfee7lghagNzFaHSOPYRRmHvIPF5+cqKiqwfPlyrF+/HqNGjcInn3yC22+/XepYRH5JqxHnW6JWwycsvkwpdQByz7///W9Mnz4dAwYMgN1ux4EDB7Bp0yaWHpEHdF21CFF79m1Ro1ZCF8UDtn0Zi8+PCIKAHTt2YNSoUbjvvvug0+lw/PhxvP766+jVq5fU8Yj83oSkaI/HEABMSPR8HPIePur0Aw6HA++99x4yMzNRX1+PRYsW4cMPP0RICJdLE4mpS8cQjOgX6f4+PjQez8etDL6NxefDLBYL1q9fj2XLlqFnz554/vnnMXr0aCiVnKgTectcfSwKjlW7dXKLy25DD4sJADew+zJ+B/VBlZWVSE9PR0xMDAoKCpCdnY1du3bhgQceYOkReVlCj3Ckp+gQGtS2f2uhQUrMHRqFTStewvTp03Hx4kUvJSRP8buoDzGZTJg5cybi4+Nx4cIFFBUVIScnB4MHD5Y6GpGsTB0Sg/SUeIQGqXCtHUEKReMZnekp8fjd2F+hpKQESqUSgwYNwtdff90+galNeHKLxARBwJdffgmDwYDCwkLMmTMHc+fORWRkpNTRiGTPkyu3NmzYgEWLFuHll1/G448/zj21PoTFJxGn04mPPvoIBoMBVVVVWLhwIWbMmOHxHWJEJD53r9w6dOgQUlNTMXDgQKxZs6Zdb42nq2PxtbOGhgZs2LABS5cuRUREBNLS0jB27FioVCqpoxGRF9TX12PevHkoKCjAe++9hwEDBkgdSfZYfO2kuroaq1atwurVqzF48GAsWrQId911Fx9/EMlEVlYWFixYgIyMDMycOZP/9iXE4vuZaosNOaUVMFWaYbY6oNWooeuqxcQk924RP378OF577TVs3rwZ48ePx7PPPov4+HgvJCciX2cymZCamor+/fvjrbfe4m0pEmHx/ajsVC1WGcuRf7QKAGC7wpvY+rhIzBkRi4Qe4dccr7i4GAaDAXl5eXjqqafwzDPPICoqykvpichfNDQ0YP78+cjLy0N2djaPGZQAiw9AVtFJZGwzwepwtnpag0IBaNQqpKfoMHVIzGV/7nK5sG3bNhgMBpw8eRILFizAE088gbAwnttHRC1t2bIF8+bNw5///GfMnj2bjz7bkeyLr7H0DqOhDVcuhwYpkZ4S31x+NpsNmzZtQmZmJkJCQpCWloaJEyfyDjwiatXRo0eRmpqKfv36Ye3atejUqZPUkWRB1sVXdqoWk9YWuXU0UWiQCusm90fBR5uwYsUK3HbbbUhLS8M999zDn9yI6LpZrVYsXLgQn3/+ObKzs5GUlCR1pIAn6+J7auM+tw+jheCC/UQp7g4+jkWLFiEhIUH0fEQkH++++y7mzp2L5557DnPnzuUP0F4k2+Krttgw9JWdLRaxtFWwSoHCxffwJHYiEkV5eTkefvhhxMTEYP369QgPD5c6UkCS7VmdOaUVHo+hVCiQs9/zcYiIACA2NhZ79uxBt27dkJiYiJKSEqkjBSTZFp+p0uzRbA9oPLfPdKZOpEREREBISAjeeOMNGAwGjB49GsuXL4dMH8x5jWyLz2x1iDSOXZRxiIh+avz48SgqKkJWVhbGjRuHCxcuSB0pYMi2+LQace7g1Wq4ZYGIvKN3797YvXs3evXqhcTEROzdu1fqSAFBtsWn66pFiNqzL1+jVkIXxc3pROQ9ISEheP3117Fs2TI88MADeO211/jo00OyLb4JSdEejyEAmJDo+ThERNcyZswYFBcXIzs7Gw899BDOnz8vdSS/Jdvi69IxBCP6RV7zduWrUSgaL6DkVgYiai8xMTEoKChA3759cfvtt6OwsFDqSH5JtsUHAHP1sdCo3bsHT6NWYY4+VuREREStCw4OxtKlS/HGG29gzJgxMBgMcLk8W6EuN7IuvoQe4UhP0SFI0bbn5Y1ndeowIDrcO8GIiK7hwQcfRHFxMd5//3088MADqK6uljqS35B18QFA/5ALsBS8gxCV4pqPPRWKxjM6f3pANRGRVHr16oVdu3ahf//+SExMxO7du6WO5Bdke2QZANTU1GDQoEF45ZVXoLvzXqw2liPvSBUUaNyc3qTpPr7kuEjM0cdypkdEPuef//wnnnjiCcyfPx+/+93voFTKfl5zVbItPqfTiVGjRiEhIQEGg6H592ssNuTsr4DpTB3MVju0miDoosIwIdG9G9iJiNrLqVOnMHnyZISFheGdd95BZGSk1JF8kmyLb/Hixdi3bx9yc3OhVouzmZ2ISGp2ux1LlixBVlYWNm3ahOHDh0sdyefIsvhycnKwaNEi7Nu3D126dJE6DhGR6HJzc/HYY4/h6aefxh/+8Ac++vwJ2RXfN998A71ej9zcXF74SEQB7fvvv8fkyZOh0WiwceNG3HzzzVJH8gmy+hGgtrYWY8eORWZmJkuPiAJe9+7dsXPnTvzyl79EYmIi8vLypI7kE2Qz43O5XHjooYfQq1cvrFy5Uuo4RETtavv27Zg+fTpmz56N9PR0qFTuHd4RCGRTfC+88AK2b9+OnTt3Ijg4WOo4RETt7vTp05gyZQqUSiU2bdqErl27Sh1JErJ41Pnpp59i7dq1yMnJYekRkWx169YNX3zxBYYNG4bExETs2LFD6kiSCPgZ37FjxzB06FB89NFH+NWvfiV1HCIin7Bjxw5MmzYNTz75JJYsWSKrR58BXXwWiwWDBw/GM888g9mzZ0sdh4jIp1RWVmLKlClwuVzYtGkTunXrdsWPq7bYkFNaAVOlGWarA1qNGrquWkxM8s+DPQK2+ARBQGpqKrRaLdatWweFu/cPEREFMKfTiYyMDLz55pvYsGED7r333uY/KztVi1XGcuQfrQIA2K5wlKM+LhJzRsQioUd4Oyd3X8AW36uvvoqcnBzs2rULGo1G6jhERD7NaDRi6tSpmDFjBp5//nls3VeBjG0mWB1OtNYSCkXjNW3pKTq/Obw/IItv+/btmDZtGoqLi9GjRw+p4xAR+YWzZ8/i0UcfRVUnHRp098PquP56aLyuzT9urgm44jt58iSGDBmCrVu3Qq/XSx2HiMivHPjPeUx8czccaPtil9AgFbKfGuLzN9gE1HaGhoYGjBs3DosXL2bpERG54c38b+FUuLfC0+pwYrWxXORE4guY4hMEAbNmzUJ8fDx++9vfSh2HiMjvVFtsyD9a1ep7eq0RBCDvSBVqLDZxg4ksYIpv5cqVOHjwINauXcsVnEREbsgprfB4DAWAnP2ej+NNAXERXUFBAV588UUUFhbihhtukDoOEZFfMlWaW2xZcIfV4YLpTJ1IibzD72d833//PR5++GG888476N27t9RxiIj8ltnqEGkcuyjjeItfF5/NZsP48ePxzDPP4L777pM6DhGRX9NqxHkIqNUEiTKOt/h18c2bNw/du3fH4sWLpY5CROT3dF21CFF7VgsatRK6qDCREnmH3xbfunXrUFBQgLfffpuLWYiIRDAhKdrjMQQAExI9H8eb/LL49u7diz/+8Y/44IMPEBbm2z9ZEBH5iy4dQzCiXyTcnUsoFEByXKTPH1ztd8V39uxZTJw4EevWrUNcXJzUcYiIAspcfSw0avc2sGvUKszRx4qcSHx+VXx2ux2pqamYMWMGHnzwQanjEBEFnIQe4UhP0SE0qG310HhWp87njysD/Oyszvnz5+PYsWP45JNPoFT6VWcTEfmVrKKTAXs7g99sYM/KysKnn36KkpISlh4RkZdNHRKDAdHhWG0sR96RKijQuDm9iQouCIKAe/t3wxx9rF/M9Jr4xYzvq6++wsiRI5GXl4df/OIXUschIpKVGosNOfsrYDpTB7PVDq0mCN06CPjLzIdwqvwwOnToIHXENvH54qupqcEdd9yBl19+GampqVLHISKiH40ePRqTJk3Co48+KnWUNvHpZ4ZOpxOPPPIIxo8fz9IjIvIx06dPx4YNG6SO0WY+PeP7wx/+gJKSEuTm5kKt9pu3I4mIZMFqtaJ79+44cOAAevbsKXWc6+azM76cnBxs2bIFW7duZekREfkgjUaD1NRUbNy4UeoobeKTM75vvvkGer0eubm5SEpKkjoOERFdRVFREaZNm4YjR474zfGRPjfjq62txdixY5GZmcnSIyLycYMHD4ZCoUBRUZHUUa6bTxWfy+XCo48+invvvRfTp0+XOg4REV2DQqHAjBkz8Pbbb0sd5br51KPOF154Adu3b8fOnTsRHBwsdRwiIroOFRUVGDBgAL7//nuEhoZKHeeafGbG9+mnn2Lt2rXIyclh6RER+ZHo6GgMGjQIH3/8sdRRrotPFN+xY8fw+OOP47333kPXrl2ljkNERG00ffp0v3nc6fVHndUWG3JKK2CqNMNsdUCrUUPXVYuJSdHo3DEEFosFgwcPxjPPPIPZs2d7MwoREXlJfX09unfvjm+++QbdunWTOk6rvFZ8ZadqscpYjvyjVQAA208ON9WolRAAjOgXiVOfrUO3EDvWrVvnN0thiYjock8++STi4uKQlpYmdZRWeaX4rvc6C0AAnHY89+BteGyY719eSEREV1dQUIDZs2fj66+/9umJjOjv8TWW3mE02K9VegCgAFTBePXzY8gqOil2FCIiakfDhg2D1WpFaWmp1FFaJWrxlZ2qRcY2Exrsrmt/8E802F3I2GbCwYpaMeMQEVE7UigUfnFwtajFt8pYDqvD6dZrrQ4nVhvLxYxDRETtbNq0adi6dStsNpvUUa5KtOKrttiQf7TqOh5vXpkgAHlHqlBj8d2/LCIial1MTAz69++Pf/7zn1JHuSrRii+ntMLjMRQAcvZ7Pg4REUlnxowZPv24U7TiM1WaW2xZcIfV4YLpTJ1IiYiISArjx49Hfn4+zp07J3WUKxKt+MxWh0jj2EUZh4iIpBEWFoYHH3wQmzdvljrKFYlWfFqNOJfFajVBooxDRETS8eXHnaIVn66rFiFqz4bTqJXQRYWJlIiIiKSi1+tx/vx5lJWVSR3lMqIV34SkaI/HEABMSPR8HCIikpZSqcSjjz7qk7M+0YqvS8cQjOgXCXdPqVEogOS4SHTuGCJWJCIiktD06dOxadMm2O2+tXZD1A3sc/Wx0KhVbr1Wo1Zhjp7ndRIRBYq+ffsiNjYWubm5UkdpQdTiS+gRjvQUHUKD2jZsaJAS6Sk6DIgOFzMOERFJzBePMJP0dgaFonGml56iw9QhMWLHICIiidXW1iImJgbHjx9H586dpY4DwIv38R2sqMVqYznyjlRBgcbN6U2a7uNLjovEHH0sZ3pERAFs8uTJGDZsGObOnSt1FADtcAN7jcWGnP0VMJ2pw6b33seUieOgiwrDhMRoLmQhIpKB3NxcLFmyBMXFxVJHAdAOxdfikykUaMdPR0REPsDpdKJnz57Yvn07br31VqnjiH8RLRER0U+pVCpMnTrVZxa5cMZHRERed+jQIYwcORL/+c9/oFK5t+1NLJzxERGR1916663o3r07tm/fLnUUFh8REbUPX9nTx0edRETULmpqatCnTx+cPHkS4eHhkuXgjI+IiNpF586d8etf/xrvvvuupDlYfERE1G584XEnH3USEVG7sdvt6NGjBwoKCtC3b19JMnDGR0RE7SYoKAiPPPKIpLM+zviIiKhdlZWV4cEHH8SJEyegVLb//IszPiIialcJCQmIiIiA0WiU5POz+IiIqN1Nnz4db7/9tiSfm486iYio3Z07dw79+vXDV6bjyD1SC1OlGWarA1qNGrquWkxM8t4NPiw+IiJqd2WnavHoy1mo7xQDlUoF2xXubNXHRWLOiFgk9AgX9XOz+IiIqF1lFZ1ExjYTrHYHBCiu+nEKBaBRq5CeosPUITGifX61aCMRERFdQ2PpHUaD3QW0UnoAIAhAg92JjG2HAUC08uPiFiIiahdlp2qRsc30Y+ldvwa7CxnbTDhYUStKDhYfERG1i1XGclgdTrdea3U4sdpYLkoOFh8REXldtcWG/KNVcHeZhyAAeUeqUGOxeZyFxUdERF6XU1rh8RgKADn7PR+HxUdERF5nqjS32LLgDqvDBdOZOo+zsPiIiMjrzFaHSOPYPR6DxUdERF6n1Yize06rCfJ4DBYfERF5na6rFiFqzypHo1ZCFxXmcRYWHxERed2EpGiPxxAATEj0fBwWHxEReV2XjiEY0S8SitYPa7kqhQJIjosU5eBqFh8REbWLufpYaNQqt16rUaswRx8rSg4WHxERtYuEHuFIT9EhNKht1RMapER6ig4DosNFycFDqomIqN00HTSdsc0Eq8PZ6kku3rqdgdcSERFRuztYUYvVxnLkHamCAo2b05s03ceXHBeJOfpY0WZ6TVh8REQkmRqLDTn7K2A6Uwez1Q6tJgi6qDBMSOQN7ERERKLg4hYiIpIVFh8REckKi4+IiGSFxUdERLLC4iMiIllh8RERkayw+IiISFZYfEREJCssPiIikhUWHxERyQqLj4iIZMXr1xJVW2zIKa2AqdKMyPF/wvzsA9B11WJikvcOICUiIroarx1SXXaqFquM5cg/WgUAsF3hygl9XCTmjIhFQo9wb0QgIiK6jFeKL6vopKSXDBIREV2N6I86G0vvMBrsrmt+rCAADXYnMrYdBgCWHxEReZ2oi1vKTtUiY5vpukrvpxrsLmRsM+FgRa2YcYiIiC4javGtMpbD6nC69Vqrw4nVxnIx4xAREV1GtOKrttiQf7Sq1ff0WiMIQN6RKtRYbGJFIiIiuoxoxZdTWuHxGAoAOfs9H4eIiOhqRCs+U6W5xZYFd1gdLpjO1ImUiIiI6HKiFZ/Z6hBpHLso4xAREV2JaMWn1YizM0KrCRJlHCIioisRrfh0XbUIUXs2nEathC4qTKRERERElxOt+CYkRXs8hgBgQqLn4xAREV2NaMXXpWMIRvSLhELh3usVCiA5LpIHVxMRkVeJuoF9rj4WGrXKrddq1CrM0ceKGYeIiOgyohZfQo9wpKfoEBrUtmFDg5RIT9FhQHS4mHGIiIguI/pFtFOHxCA9JR6hQaprPvZUABDsNsxI0PKAaiIiahdeu4/vYEUtVhvLkXekCgo0bk5v0nQfX3JcJPrYjmPNS39ESUkJIiIivBGFiIiomdeKr0mNxYac/RUwnamD2WqHVhMEXVQYJiT+9wb2hQsX4vDhw/j000+hUrn3HiEREdH18HrxXQ+Hw4GRI0di6NChePHFF6WOQ0REAUz09/jcoVarkZ2djY0bN+LDDz+UOg4REQUwn5jxNSkpKcHo0aOxa9cu6HQ6qeMQEVEA8okZX5M77rgDL7/8MsaMGQOz2Sx1HCIiCkA+NeNr8pvf/AaVlZX4xz/+AaXSp7qZiIj8nE+2yvLly3H27Fn85S9/kToKEREFGJ+c8QHA6dOncccdd2DdunUYNWqU1HGIiChA+GzxAcDu3bsxfvx47NmzB3369JE6DhERBQCffNTZZNiwYViyZAnGjh2LixcvSh2HiIgCgE/P+ABAEAQ89thjsNls2Lx5MxTu3ntEREQEH5/xAYBCocCbb76Jo0ePYtmyZVLHISIiP+fzM74m3333HYYMGYLNmzcjOTlZ6jhERLJRbbEhp7QCpkozzFYHtBo1dF21mJgU7ZeXh/tN8QHAzp07MWXKFOzduxc9e/aUOg4RUUArO1WLVcZy5B+tAgDYrnDLjj4uEnNGxCKhR7g0Id3gV8UHAJmZmdi6dSsKCgoQGhoqdRwiooCUVXQSGdtMsDqcaK0lFApAo1YhPUXnN/eq+l3xCYKAyZMnIzQ0FH/729+42IWISGSNpXcYDXbXtT/4R6FBSqSnxPtF+fn84pafUygUWL9+Pfbt24c333xT6jhERAGl7FQtMraZ2lR6ANBgdyFjmwkHK2q9E0xEfld8ANChQwd88MEHeP755/Hll19KHYeIKGCsMpbD6nC69Vqrw4nVxnKRE4nPL4sPAGJjY/H2228jNTUVp0+fljoOEZHfq7bYkH+0qtX39FojCEDekSrUWGziBhOZ3xYfAKSkpGD27NmYOHEiLl26JHUcIiK/llNa4fEYCgA5+z0fx5v8uvgAID09HV26dMH8+fOljkJE5NdMleYWWxbcYXW4YDpTJ1Ii7/D74lMqlXjnnXewY8cO/P3vf5c6DhGR3zJbHSKNYxdlHG9RSx1ADJ06dcKHH36I4cOH47bbbsOgQYOkjkRE5He0GnEqQasJEmUcb/H7GV+T+Ph4vPXWWxg/fjzOnTsndRwiIr+j66pFiNqzWtColdBFhYmUyDsCpvgAYNy4cZgyZQoefvhhOBziTNmJiORiQlK0x2MIACYkej6ONwVU8QHA//3f/yEkJAS///3vpY5CRORXunQMwe1dQwCXewtcFAogOS7S5w+uDrjiU6lU2Lx5Mz744ANs2bJF6jhERH5BEASsX78eeSsXI0jl3lGQGrUKc/SxIicTX0Asbvm5iIgIfPDBB/j1r3+N/v37Y8CAAVJHIiLyWXV1dfjNb36Dr776Csb330PpD6FuntWpw4DocO8FFUnAzfiaJCQkYPny5Rg7dizOnz8vdRwiIp9UVlaGQYMGQaPRoLi4GPHxjQdNp6fEIzRIhWvdA6BQAKFBKr85oBrww9sZ2mrBggUwmUz49NNPoVKppI5DROQTBEHA2rVrkZ6ejtdffx1Tpky57GMOVtRitbEceUeqoEDj5vQmTffxJcdFYo4+1i9mek0CvvjsdjtGjhyJYcOG4cUXX5Q6DhGR5MxmM2bNmoVDhw7h3XffRVxcXKsfX2OxIWd/BUxn6mC22qHVBEEXFYYJibyB3WedO3cOgwYNwooVKzBmzBip4xARSebAgQNITU3FPffcg2XLlsnyQm9ZFB8AFBcXY/To0SgoKIBOp5M6DhFRuxIEAWvWrMGSJUvwxhtvYNKkSVJHkoxsig8A1q9fD4PBgOLiYmi1WqnjEBG1ix9++AEzZ87EsWPH8O6776Jv375SR5JUwK7qvJInnngCer0e06dPh8vNDZpERP6ktLQUSUlJiIyMRGFhoexLD5BZ8QHA8uXLUVlZib/85S9SRyEi8hpBELBy5UqMGjUKL730ElatWgWNRiN1LJ8QkBvYWxMSEoJ//OMfuOOOO5CYmIhRo0ZJHYmISFS1tbV44okncPLkSRQWFqJPnz5SR/IpspvxAUC3bt2QnZ2NGTNm4Pjx41LHISISTUlJCRITE9G9e3fs2bOHpXcFslrc8nMrV67EX//6VxQWFqJDhw4AgGqLDTmlFTBVmmG2OqDVqKHrqsXEJP/cr0JE8iAIApYvX46XXnoJb775JsaPHy91JJ8l6+ITBAEzZszApUuX8PtXV2O18Tjyj1YBAGxXOKFAHxeJOSNikdAjXJrARERXcOHCBTz22GM4ffo0tm7dit69e0sdyafJuvgAoKGhAUmTFuDSraPhVCjR2t+GQtF4+nh6is5vzqQjosC2d+9eTJo0CWPGjMErr7yC4OBgqSP5PNktbvm5f5SdhXPAg3A4hMYbFFshCECD3YmMbYcBgOVHRJIRBAGvvfYaXn31Vbz11ls8laoNZF18ZadqkbHNBJujbZPeBrsLGdtMGBAd7lcHsxJRYKipqcGMGTNQVVWFvXv3IiYmRupIfkWWqzqbrDKWw+pwuvVaq8OJ1cZykRMREbVuz549SExMRFxcHHbt2sXSc4NsZ3zVFhvyj1a1+p5eawQByDtShRqLjas9icjrXC4XMjMzsXTpUqxbtw4PPPCA1JH8lmyLL6e0wuMxFABy9ldg1nDukyEi76mursa0adNQW1uLkpIS9OzZU+pIfk22jzpNleYWWxbcYXW4YDpTJ1IiIqLL7d69G4mJibjtttuQn5/P0hOBbGd8ZqtDpHHsooxDRPRTLpcLr7zyCpYvX46//e1vSElJkTpSwJBt8Wk14nzpWk2QKOMQETU5d+4cpk2bBovFgn379iE6OlrqSAFFto86dV21CFF79uVr1EroosJESkREBOTn5yMxMRGJiYkwGo0sPS+QbfFNSPL8/0x2hwN3x9wgQhoikjun04kXX3wRkyZNwvr16/HSSy9BrZbtQzmvkm3xdekYghH9IqFQuPd6BYDOtjMYPLA/5s6dy1seiMhtZ8+exf3334/t27dj3759uO+++6SOFNBkW3wAMFcfC41a5dZrNUEqrFuYikOHDiE8PBxDhgzBxIkTUVxcLHJKIgpkeXl5SExMxJAhQ7Bjxw50795d6kgBT/aHVGcVnUTGtsNosF//1obQICXSU+JbnNVpsViwfv16LFu2DL169UJaWhpSUlKgVMr6ZwsiuoqmR5tvvfUWNmzYgJEjR0odSTZkX3xAU/mZYHU4Pb6dweFw4L333oPBYIDVasWiRYswZcoUhITwdBcialRZWYkpU6bA5XJh8+bNiIqKkjqSrLD4fnSwoharjeXIO1IFBRo3pzdpuo8vOS4Sc/Sx13UwtSAI2LlzJwwGAw4ePIh58+Zh1qxZuPHGG732NRCR7/viiy8wbdo0PPXUU/jTn/4Elcq9t1vIfSy+n6mx2JCzvwKmM3UwW+3QaoKgiwrDhET3b2A/ePAgli5dik8++QTTp0/HggULePoCkcw4nU688MILWL9+PTZu3Ii7775b6kiyxeJrRxUVFc2nMNx///1IS0vDwIEDpY5FRF52+vRpPPLII1Cr1cjKykLXrl2ljiRrXHnRjqKjo2EwGPDtt99i4MCBGD16NEaOHInPP/8c/PmDKDB99tlnSEpKwj333IPPPvuMpecDOOOT0KVLl7BlyxZkZmZCpVJh0aJFePjhhxEUxGPQiHxJtcWGnNIKmCrNMFsd0GrU0HXVYmLS1d8CcTgceO6557BhwwZkZWVBr9e3b2i6KhafDxAEAbm5uTAYDCgvL8f8+fMxc+ZMhIXxODQiKZWdqsUqYznyj1YBQIsbXZoWvenjIjFnRCwSeoQ3/1lFRQUeeeQRhIaGYuPGjbjpppvaOTm1ho86fYBCocCoUaOwc+dOvP/++yguLsYtt9yCxYsX4/Tp01LHI5KlrKKTmLS2CNsPn4XN4brsGjPrj7/3+aGzmLS2CFlFJwEA//rXvzBo0CDcf//9+Ne//sXS80Gc8fmoEydOYNmyZcjKysKYMWPw7LPPon///lLHIpIFdw620AQp0f/SERRvWopNmzZh+PDhXkxInuCMz0fdcsstWLFiBY4dO4bevXvjnnvuwejRo2E0GrkQhsiLyk7VImObqU2lBwBWuwv7XTHI2raLpefjWHw+rnPnzvjf//1fnDx5Eg899BBmzZqFwYMH491334XDIc5lukT0X6uM5bA6nO69WB2ELWU14gYi0fFRp59xuVz4+OOPYTAYcObMGSxcuBCPPfYYOnToIHU0Ir9XbbFh6Cs7L3s/ry1C1Ers+f3dbh94Qd7HGZ+fUSqVGDNmDL788ktkZWVhx44duOWWW7BkyRKcO3dO6nhEfi2ntMLjMRQAcvZ7Pg55D4vPj91555344IMPsHv3bpw7dw5xcXGYPXs2jh07JnU0Ir9kqjR7NNsDGld7ms7UiZSIvIHFFwD69euHNWvW4MiRI7jpppswdOhQjBs3DoWFhVJHI/IrZqs475ubrXZRxiHvYPEFkJtuugl//vOfceLECdx9992YMmUKhg0bho8++ggul2c/xRLJQViIWpRxtBqevuTLuLglgDkcDrz//vswGAwwm8149tlnMW3aNGg0GqmjEfkEl8uFr7/+GkajEUajEXsudIBm0DhAHez2mBq1EgtG9sOs4X1ETEpiYvHJgCAIyM/Ph8FgQGlpKZ5++mnMmTMHERERUkcjalculwvffPNNc9Hl5+cjIiICer0eer0eA345FOM3HOKqzgDH4pOZb775BpmZmfjoo48wdepULFiwALfccovUsYi8wuVy4dChQy2KLjw8vLno9Ho9unfv3uI1T23ch+2Hz8Kd74wKBXDfrTdjzdRBIn0F5A0sPpn6/vvvsWLFCqxbtw4jR45EWloakpKSpI5F5BFBEC4rOq1W26LooqOjWx2j7FQtJq0tQoO97ZvYQ4NUyH5qCAZEh7v5FVB7YPHJnNlsxrp16/D6668jNjYWaWlpuP/++6FQKNo8ljtXtxB5QhAEHD58uLnojEYjwsLCWhRdjx492jyuO2d1hgYpkZ4Sj6lDYtr8+ah9sfgIAGC325GdnQ2DwQCXy4VFixZh8uTJCA6+9pv87l7dQtRWgiDAZDK1KLoOHTq0KLqePXuK8rkay88Eq8PZ6mNPhQLQqFVIT9Gx9PwEi49aEAQB27dvh8FgwOHDhzFv3jzMmjULnTp1uuLH85sDeZMgCDhy5EiLogsNDW1RdL169fLa5z9YUYvVxnLkHamCAo2b05s0/VCXHBeJOfpYPt70Iyw+uqoDBw4gMzMTubm5ePzxx/Hb3/62xfsjfBxEYhMEAUePHm1RdMHBwUhOTm4uupiYmHbPVWOxIWd/BUxn6mC22qHVBEEXFYYJiXyM749YfHRN3333HV5//XVs2LAB//M//4NFixZBuLEnFwCQxwRBwLFjx2A0GpGXlwej0YigoKAWRcdVxyQ2Fh9dtwsXLmDNmjVYsWIFOo/5Ay6G94E7/+fhkm/5EgQB5eXlLYpOpVJdVnTuLK4iul4sPmqz0zV1uGtpPpyC+9+cuMlXHgRBwPHjx1sUnUKhaFF0vXv3ZtFRuxLnYDqSlY+/Pge1SgWnB6dbNF3dwmOdAosgCPj2229bFJ0gCEhOTkZycjJeeOEF9OnTh0VHkmLxUZvx6hbfIuX+SUEQcOLEiRZF53Q6m2d0zz33HGJjY1l05FP4qJPa7PENJdhp8vzS2862MxgZXI4uXbogMjKy+dem/+7YsSO/YbZCiv2TgiDg5MmTLYrObre3eHTZt29f/u9GPo3FR202P/sAPvzqtMfj/KJjPYYqy1FVVYXq6urLfnU4HFctxSv92rlzZ6hUKhG+Qt/Xnvsnm4quqexsNluLouvXrx+LjvwKH3VSm+m6ahGirvTocadGrcQDdyVi1vCJV/2Y+vp6VFdXX7EUDxw4cNnvXbhwAZ06dbruooyMjMQNN9zg9tcglbbsnxQEoMHuRMa2wwBwXeX33XfftSg6q9XaXHKLFy9GXFwci478Gmd81GbVFhuGvrLT565ucTqdOH/+/BWLsqqq6oq/p1AorlqKVyrMiIgIKJXS3d/sjQOU//Of/7Qouvr6+hYno+h0OhYdBRQWH7klEK5uEQQB9fX1Vy3FKxVmXV0dbrzxxjY9ghXz4l8x/t7TR9zcougsFkuLoouPj2fRUUBj8ZFb5Hp1i8PhQE1NTauzyJ//WXBwcKuPW3/+e506dbrirFKMmTacdlzcshD6IUnNRXfrrbey6EhWWHzkNp7VeW2CIKCurq5NRXnx4kV07tz5slKsvPE2lLmi4YT7j1qDVQosHNkPs0fEivhVEvkXLm4htzWVF29nuDqFQgGtVgutVovevXtf12suXbqEmpqay0rRdOYGj0oPAC45BRyptHg0BpG/Y/GRR6YOicGA6HBe3SKi4OBgREVFISoqqsXv799Qgm9F2D9ptto9HoPIn7H4yGMDosOxZuogXt3iZVqNOP9ctZogUcYh8lcsPhJN544hPHvTi8TaP6mLChMxFZH/kW5DEhG1yYSk6Gt/0DUIACYkej4OkT9j8RH5iS4dQzCiXyTc3XmgUDS+18rHziR3LD4iPzJXHwuN2r3zSDVqFebouY2BiMVH5EcSeoQjPUWH0KC2/dNt3D+p46paInBxC5Hf4f5JIs/w5BYiP3Wwopb7J4ncwOIj8nPcP0nUNiw+IiKSFS5uISIiWWHxERGRrLD4iIhIVlh8REQkKyw+IiKSFRYfERHJCouPiIhkhcVHRESywuIjIiJZ+X+0lCnp6Dl6TAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "\n",
    "#calling the gfa parser first to retrieve the data from gfa\n",
    "#user input\n",
    "input_name=\"ED144/chrom2.gfa\"\n",
    "x=gfa_parser(input_name)\n",
    "edges=x.gfa_parser(input_name)[0]\n",
    "coverage=x.gfa_parser(input_name)[1]\n",
    "sequences=x.gfa_parser(input_name)[2]\n",
    "\n",
    "#using networkx package to write the edges from gfa to graph\n",
    "G=nx.Graph()\n",
    "G.add_edges_from(edges)\n",
    "#getting the first node in the graph\n",
    "n=list(G.nodes())[0]\n",
    "#dfs traversal\n",
    "successors=nx.dfs_successors(G,n)\n",
    "length=len(successors)\n",
    "print (successors)\n",
    "\n",
    "#traversing the old graph to write to a new graph, traverse through each node at least once\n",
    "new_graph=nx.Graph()\n",
    "count=0\n",
    "\n",
    "while (count < length):\n",
    "    keys=list(successors.keys())\n",
    "    key=keys[0]\n",
    "    val=successors[key]\n",
    "    print (\"key, val, count\",key, val, count)\n",
    "    #print (\"successors=\", successors)\n",
    "    \n",
    "    #adding the first edge if there is only branch down the tree\n",
    "    if (count==0) and(len(val)==1):\n",
    "        new_graph.add_edge(key, val[0]) \n",
    "        count=count+1 \n",
    "        successors.pop(key)\n",
    "    \n",
    "    #if the first node has two branches and picking the edge with the shortest depth\n",
    "    elif (count==0) and (len(val)!=1):\n",
    "        two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "        successors=two_branches[0]\n",
    "        count=count+1 \n",
    "        successors.pop(key)\n",
    "        \n",
    "    #adding the rest of the nodes \n",
    "    elif (count!=0):\n",
    "        #getting the last node\n",
    "        last_node=get_last_edge(new_graph)[1]\n",
    "        \n",
    "        #making sure the last node is the same as the next key\n",
    "        if (last_node==key):\n",
    "            \n",
    "            #if there is only one branch\n",
    "            if (len(val)==1):\n",
    "                new_graph=check_before_adding(key, val[0], new_graph)\n",
    "                count=count+1 \n",
    "                successors.pop(key)\n",
    "                \n",
    "            #if there are two branches\n",
    "            else:\n",
    "                #print (\"adding edges with more than one path\")\n",
    "                two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "                successors=two_branches[0]\n",
    "                new_graph=two_branches[1]\n",
    "                count=count+1 \n",
    "                successors.pop(key)\n",
    "   \n",
    "        \n",
    "        #if the last_node is not the same as the key, not in continuation \n",
    "        else:\n",
    "            \n",
    "            #finding the continuation node in the dict, if it exists \n",
    "            if (last_node in list(successors.keys())):\n",
    "                val=successors[last_node]\n",
    "                \n",
    "                #if there is only one branch from key\n",
    "                if (len(val)==1):\n",
    "                    new_graph=check_before_adding(last_node, val[0], new_graph)\n",
    "                    count=count+1 \n",
    "                    successors.pop(key)\n",
    "                \n",
    "                #if there is more than one branch down the tree\n",
    "                else:\n",
    "                    two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "                    successors=two_branches[0]\n",
    "                    new_graph=two_branches[1]\n",
    "                    count=count+1 \n",
    "                    successors.pop(key)\n",
    "                    \n",
    "            #finding the continuation node in the dict, if it doesn't exists \n",
    "            else:\n",
    "                #finding the shortest path to the key from the last node\n",
    "                path=path_to_first_node(last_node, key, G)\n",
    "                left_path=path[0]\n",
    "                right_path=path[1]\n",
    "                new_graph=shortest_path_to_first_node(left_path, right_path, new_graph)               \n",
    "\n",
    "    #print (\"count, length of keys, keys\", count, len(keys), keys)\n",
    "    #print (new_graph.edges())\n",
    "    print (\"successors=\", successors)\n",
    "    \n",
    "#adding last node\n",
    "last_node=get_last_edge(new_graph)[1]\n",
    "last_node=last_node.split('r')[0]\n",
    "path=path_to_first_node(last_node, n, G)\n",
    "left_path=path[0]\n",
    "right_path=path[1]\n",
    "new_graph=shortest_path_to_first_node(left_path, right_path, new_graph)               \n",
    "print (new_graph.edges())\n",
    "\n",
    "#print (new_graph.edges)\n",
    "nx.draw_circular(new_graph)\n",
    "\n",
    "# writing the graph to fasta file \n",
    "final=graph_to_fasta(new_graph, input_name, sequences)\n",
    "SeqIO.write(final, \"ED144/ED144-chrom2-final-redone.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notes \n",
    "\n",
    "The last shell may have to be changed to functions as well.. Not sure how yet in a Jupyter notebook. If changed to a code, then add argparse python package. \n",
    "\n",
    "Replace the first line - name=\"chrom1.gfa\" and last line \"ED252-chrom1-final.fasta\". Replace the names for different files\n",
    "\n",
    "This generated two files, one for chromosome 1 and another for chromosome 2. In bash ran the command, \n",
    "    \n",
    "`cat ED144-chrom1-final.fasta ED144-chrom2-final.fasta >ED144-final.fasta`\n",
    "    \n",
    "   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
